### ISSUE TITLE
Resolving the monad instance for `do` notation

### ISSUE THREAD

**Issue by berndlosert** [agda/agda#4339]
It seems that Agda cannot figure out which Number instance to use inside a do block:

```agda
 open import Agda.Builtin.Unit
 open import Agda.Builtin.Nat
 open import Agda.Builtin.FromNat
 open import Agda.Builtin.Int
 open import Agda.Builtin.List

 instance
   Number:Nat = record {
       Constraint = \ _ -> ⊤;
       fromNat = \ n -> n
     }

   Number:Int = record {
       Constraint = \ _ -> ⊤;
       fromNat = \ n -> pos n
     }

 record Monad (M : Set -> Set) : Set1 where
    infixl 1 _>>=_ _>>_
    field
      return : forall {X} -> X -> M X
      _>>=_ : forall {X Y} -> M X -> (X -> M Y) -> M Y

    _>>_ : forall {X Y} -> M X -> M Y -> M Y
    x >> y = x >>= (\ _ -> y)

 open Monad {{...}}

 instance
   Monad:List : Monad List
   Monad:List .return x = x ∷ []
   Monad:List ._>>=_ [] k = []
   Monad:List ._>>=_ (x ∷ xs) k = k x ++ (xs >>= k)
     where
       _++_ : forall {X} -> List X -> List X -> List X
       [] ++ xs = xs
       (x ∷ xs) ++ ys = x ∷ (xs ++ ys)

 test1 : List Int
 test1 = 1 ∷ [] >>= \ x -> return {List} x -- Note the explicit {List}
 
 -- Will not compile unless I comment out the Number:Nat instance. 
 test2 : List Int
 test2 = -- equivalent to: 1 ∷ [] >>= \ z -> return z 
   do
     x <- 1 ∷ []
     return x
```

If I don't comment out the`Number:Nat` instance, I get the following error:

```
Checking foo (/private/tmp/foo.agda).
Failed to solve the following constraints:
  Resolve instance argument
    _r_91 : (x1 : _X_80) -> Monad (_M_90 (x = x1))
  No candidates yet
  Resolve instance argument _87 : Number.Constraint _r_86 1
  No candidates yet
  Resolve instance argument _r_86 : Number _X_80
  Candidates
    record { Constraint = \ _ -> ⊤ ; fromNat = \ n {{_}} -> n } :
      Number Nat
    record { Constraint = \ _ -> ⊤ ; fromNat = \ n {{_}} -> pos n } :
      Number Int
  [118] _M_90 _X_80 =< List Int : Set
  _93 := \ x -> Monad.return (_r_91 (x = x)) x
    [blocked on problem 118]
Unsolved metas at the following locations:
  /private/tmp/foo.agda:46,7-9
  /private/tmp/foo.agda:46,10-11
  /private/tmp/foo.agda:47,5-11
  /private/tmp/foo.agda:47,5-13

shell returned 1
```

**Comment by jespercockx** [comment]
Thanks for reporting this issue.

The problem is that Agda cannot figure out which `Monad` instance to use. If it could, then the constraint `_M_90 _X_80 =< List Int` would be solved (with `_X_80 := Int`) and the right `Number` instance would be found as well.

The underlying problem seems to be that Agda fails to resolve the instance argument for `return x` because the variable `x : _X_80` is in scope, and Agda isn't sure that solving `_X_80` will not bring additional instance candidates in scope (see https://github.com/agda/agda/issues/3358 for why this is important).

On one hand, Agda is actually behaving as expected, but on the other hand from a user perspective it is quite bad that Agda cannot resolve the instances in this case. So I'm not really sure what to do here.

**Comment by berndlosert** [comment]
> on the other hand from a user perspective it is quite bad that Agda cannot resolve the instances in this case.

Indeed. This makes working with Agda less ergnomic since I have to be careful about what Number instances are in scope. There are three ways that I have discovered to work around the problem when multiple Number instances are in scope:

- Write `return {List}` instead of just `return`.
- Define the lists you need in the do block outside the do block.
- Write the lists like this:`(id {Int} 1) ∷ 2  ∷ []`; or like this: `id {List Int} (1 ∷ 2  ∷ [])`.

**Comment by gallais** [comment]
> Write the lists like this:(id {Int} 1) ∷ 2 ∷ []; or like this: id {List Int} (1 ∷ 2 ∷ []).

You should probably use something like [`_∋_`](https://agda.github.io/agda-stdlib/Function.Base.html#4269) instead.

**Comment by jespercockx** [comment]
It's not clear what we can do to improve on the current situation, so I'm closing this issue for now. Concrete proposals to enhance instance search are definitely still welcome.

**Comment by gallais** [comment]
> It's not clear what we can do to improve on the current situation,
  so I'm closing this issue for now. Concrete proposals to enhance
  instance search are definitely still welcome.

In Idris2 we have qualified `do`s: `M.do` means that the `do` block
should be desugared using `M.(>>=)` and `M.>>`.

Another popular solution is to have `using` as a way to restrict the
candidates for resolution: by writing `using [namespace] t`, the
expression `t` will preferentially resolve definitions in the provided
namespace rather than the full scope.

**Comment by andreasabel** [comment]
Being an `instance`-skeptic, I append `where open M` to my functions using do-notation, where `M` is the module defining the monad I want to use.  Qualified `do` would make this less noisy, and also allow different monads with `do` in the same function.

**Comment by UlfNorell** [comment]
I'm not a fan of reopening and repurposing old issues. To figure out that this issues is (now) about qualified `do` you have to scroll down pretty far, through a bunch of discussion on instance resolution. Please open a fresh issue and link to this one for context.
